### PR TITLE
dataconnect(chore): add `last_update_sequence_number` to cache db

### PR DESCRIPTION
**Pull request by dconeybe**
This pull request introduces a new `last_update_sequence_number` column to the Data Connect SQLite cache database for queries and entities. This sequence number tracks the order of cache updates and will be used by subsequent work to avoid publishing duplicate query results from cache in realtime query subscriptions.

### Highlights

* **Monotonic Sequence Tracking**: Configured `DataConnectCacheDatabase` to generate monotonically increasing sequence numbers using a sequence helper table, assign them to queries and entities during update transactions, and report the maximum sequence number across a query and all of its referenced entities upon retrieval.
* **Incremental Database Schema Migrator**: Upgraded the cache database schema to version `1.1.0` and structured the schema migrator to perform incremental multi-step migrations, starting with database creation at `1.0.0` and applying the migration to `1.1.0` which adds the `last_update_sequence_number` columns.
* **Lexicographical Byte Array Ordering**: Conformed `ImmutableByteArray` to the `Comparable` interface, supported by a backport of signed lexicographical byte array comparison (`contentCompareTo`) to ensure compatibility with Android versions below API level 33.
* **Comprehensive Integration and Unit Tests**: Added regression tests for database migration and verification of database state, along with thorough testing of correct sequence number calculation and stable byte array sorting.

<details>
<summary><b>Changelog</b></summary>

* **DataConnectCacheDatabase.kt**
  * Added `last_update_sequence_number` tracking logic, including the `SqliteSequenceNumber` inline value class, sequence number generation helper, and maximum last update sequence number computation during query rehydration.
* **DataConnectCacheDatabaseMigrator.kt**
  * Upgraded the database schema version to `1.1.0`, introduced an incremental migration runner loop, and implemented step migration from `1.0.0` to `1.1.0` to alter `queries` and `entities` tables.
* **ByteArrayUtils.kt**
  * Created a utility file that backports signed lexicographical byte array comparison (`contentCompareTo`) for Android devices running on API levels below 33.
* **ImmutableByteArray.kt**
  * Conformed `ImmutableByteArray` to the `Comparable<ImmutableByteArray>` interface, utilizing the new `contentCompareTo` logic to achieve a stable cross-process sort order.
* **DataConnectCacheDatabaseMigratorUnitTest.kt**
  * Added integration tests verifying that existing data is preserved during progressive `1.0.0` to `1.1.0` migrations, and that proper exceptions are thrown when the database version or application ID is unrecognized.
* **DataConnectCacheDatabaseUnitTest.kt**
  * Added unit tests verifying proper sequence number updates during query insertion and verifying sequence number retrieval logic, including edge cases with null database values.
* **ImmutableByteArrayUnitTest.kt**
  * Added property and unit tests for the newly introduced comparison behaviors of `ImmutableByteArray`, covering identical arrays, prefixes, and different array lengths.
* **CartesianProduct.kt**
  * Added a utility function to compute the cartesian product of two iterables to support unit testing.
* **TestValueGenerator.kt**
  * Added a utility class to facilitate deterministic test sample generation from Kotest arbitrary data generators (`Arb`) outside of standard property-based tests.

</details>

### CODE REVIEW

File: firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabaseMigrator.kt [hunk 1]
```diff
@@ -34,18 +35,102 @@ private constructor(private val sqliteDatabase: SQLiteDatabase, private val logg
...
   }
 
-  private fun migrate() {
+  private fun migrate(stopAtVersionForTesting: SemanticVersion? = null) {
+    if (stopAtVersionForTesting != null) {
+      logger.warn {
+        "WARNING q9dfsdxdmp: migrate() called with " +
+          "" +
+          "stopAtVersionForTesting=$stopAtVersionForTesting, " +
+          "which should ONLY be done by tests for the SDK itself"
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabaseMigrator.kt:61]: Done in 44aee3b453fc42d68a6df3f1a465ebbabe925ef5

File: firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabaseMigrator.kt [hunk 2]
```diff
@@ -34,18 +35,102 @@ private constructor(private val sqliteDatabase: SQLiteDatabase, private val logg
...
     logger.debug { "migrate() completed" }
   }
 
+  private fun migrateDatabase(stopAtVersionForTesting: SemanticVersion?) {
+    if (stopAtVersionForTesting != null) {
+      logger.warn {
+        "WARNING n7d2qazseq: migrateDatabase() called with " +
+          "" +
+          "stopAtVersionForTesting=$stopAtVersionForTesting, " +
+          "which should ONLY be done by tests for the SDK itself"
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabaseMigrator.kt:75]: Done in 44aee3b453fc42d68a6df3f1a465ebbabe925ef5

File: firebase-dataconnect/testutil/src/main/kotlin/com/google/firebase/dataconnect/testutil/CartesianProduct.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.firebase.dataconnect.testutil
+
+fun <T1, T2> cartesianProduct(
```
Comment (aashishpatil-g) [inline firebase-dataconnect/testutil/src/main/kotlin/com/google/firebase/dataconnect/testutil/CartesianProduct.kt:18]: Where is this used?

**Review by aashishpatil-g — APPROVED** [review]

File: firebase-dataconnect/testutil/src/main/kotlin/com/google/firebase/dataconnect/testutil/CartesianProduct.kt (hunk 3 above)
Comment (dconeybe) [reply firebase-dataconnect/testutil/src/main/kotlin/com/google/firebase/dataconnect/testutil/CartesianProduct.kt:18]: The `cartesianProduct()` function is used in the unit tests in the file `DataConnectCacheDatabaseMigratorUnitTest.kt` in 2 places:

1. https://github.com/firebase/firebase-android-sdk/blob/2a36318939a5fbfe47735a9a2e659a8e9d5d77f1/firebase-dataconnect/src/test/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabaseMigratorUnitTest.kt#L551
2. https://github.com/firebase/firebase-android-sdk/blob/2a36318939a5fbfe47735a9a2e659a8e9d5d77f1/firebase-dataconnect/src/test/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabaseMigratorUnitTest.kt#L573
